geom/type_polygon.go

```diff
@@ -592,3 +592,40 @@ func (p Polygon) Summary() string {
 func (p Polygon) String() string {
 	return p.Summary()
 }
+
+// Simplify returns a simplified version of the Polygon by applying the
+// Ramer-Douglas-Peucker algorithm to each constituent ring. If the exterior
+// ring collapses to a point or single linear element, the empty Polygon is
+// returned. If any interior ring collapses to a point or a single linear
+// element, then it is omitted from the final output. The output Polygon will
+// be invalid if any rings in the input become non-rings (e.g. via self
+// intersection) in the output, or if any two rings were to interact in ways
+// prohibited by Polygon validation rules (such as intersecting at more than
+// one point). In these cases, an error is returned. Construction behaviour of
+// the output (which includes omitting errors) may be controlled via
+// ConstructorOptions.
+func (p Polygon) Simplify(threshold float64, opts ...ConstructorOption) (Polygon, error) {
+	exterior := p.ExteriorRing().Simplify(threshold)
+
+	// If we don't have at least 4 coordinates, then we can't form a ring, and
+	// the polygon has collapsed either to a point or a single linear element.
+	// Both cases are represented by an empty polygon.
+	hasCollapsed := func(ring LineString) bool {
+		return ring.Coordinates().Length() < 4
+	}
+	if hasCollapsed(exterior) {
+		return Polygon{}, nil
+	}
+
+	n := p.NumInteriorRings()
+	rings := make([]LineString, 0, n+1)
+	rings = append(rings, exterior)
+	for i := 0; i < n; i++ {
+		interior := p.InteriorRingN(i).Simplify(threshold)
+		if !hasCollapsed(interior) {
```

Comment:
The bug fix mentioned in the PR description is here. Previously, the condition on this if statement was `!interior.IsRing` and was changed to `!hasCollapsed`. The implication of the bug was that if an interior ring became invalid but _didn't_ collapse, we would omit it from the output. Instead, we now give an error (which matches GEOS and PostGIS behaviour).